### PR TITLE
Add room database retrieval and validation

### PR DESCRIPTION
**Pull request by peppy**
- [x] Depends on https://github.com/ppy/osu-server-spectator/pull/3

Note that this currently breaks the sample client project (or, rather, it requires a database for it to function correctly). I'm figuring out the best way to fix this, so while this PR is usable for testing against an actual database backing, don't merge just yet.

### CODE REVIEW

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs [hunk 1]
```diff
@@ -117,6 +153,9 @@ public async Task LeaveRoom()
                         active_rooms.Remove(room.RoomID);
                     }
 
+                    using (var conn = Database.GetConnection())
+                        await conn.ExecuteAsync("UPDATE multiplayer_rooms SET ends_at = NOW() WHERE id = @RoomID", new { room.RoomID });
```
Comment (smoogipoo) [inline osu.Server.Spectator/Hubs/MultiplayerHub.cs:157]: I don't like this way of creating the object - it's too magic for my liking and relying on what I can only assume is .NET property ordering. Would you be fine if I changed these to be implemented like `new { RoomID = room.RoomID })`?

The official docs do this: https://dapper-tutorial.net/async

The difficulty calculator does this:
https://github.com/ppy/osu-difficulty-calculator/blob/0e4bc6bf8ca1536636110a8a4b10c8ac8674743b/osu.Server.DifficultyCalculator/ServerDifficultyCalculator.cs#L81-L91

Comment (smoogipoo) [reply osu.Server.Spectator/Hubs/MultiplayerHub.cs]: Made the changes anyway. As it is it can/will fail if the name of any of these properties change.

**Review by smoogipoo — APPROVED** [review]

File: osu.Server.Spectator.Tests/MultiplayerFlowTests.cs [hunk 2]
```diff
@@ -510,6 +510,15 @@ public TestMultiplayerHub(MemoryDistributedCache cache)
             {
             }
 
+            protected override Task<MultiplayerRoom> RetrieveRoom(long roomId)
```
Comment (smoogipoo) [inline osu.Server.Spectator.Tests/MultiplayerFlowTests.cs:513]: This is bypassed here, but the rest of the methods still fail unfortunately. For example:

```
osu.Server.Spectator.Tests.MultiplayerFlowTests.AllUsersBackingOutFromLoadCancelsTransitionToPlay

MySqlConnector.MySqlException: Unable to connect to any of the specified MySQL hosts.

MySqlConnector.MySqlException
Unable to connect to any of the specified MySQL hosts.
   at MySqlConnector.Core.ServerSession.ConnectAsync(ConnectionSettings cs, Int32 startTickCount, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 381
   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 110
   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ConnectionPool.cs:line 140
   at MySqlConnector.MySqlConnection.CreateSessionAsync(ConnectionPool pool, Int32 startTickCount, Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlConnection.cs:line 846
   at MySqlConnector.MySqlConnection.OpenAsync(Nullable`1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlConnection.cs:line 410
   at MySqlConnector.MySqlConnection.Open() in /_/src/MySqlConnector/MySqlConnection.cs:line 376
   at osu.Server.Spectator.Database.GetConnection() in /home/smgi/Repos/osu-server-spectator/osu.Server.Spectator/Database.cs:line 17
   at osu.Server.Spectator.Hubs.MultiplayerHub.updateDatabaseParticipants(MultiplayerRoom room) in /home/smgi/Repos/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerHub.cs:line 320
   at osu.Server.Spectator.Hubs.MultiplayerHub.JoinRoom(Int64 roomId) in /home/smgi/Repos/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerHub.cs:line 85
```

Seems like perhaps the rest of the code needs to be delegated?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Server.Spectator.Tests/MultiplayerFlowTests.cs (hunk 2 above)
Comment (peppy) [reply osu.Server.Spectator.Tests/MultiplayerFlowTests.cs:513]: Seems like we need CI on this project..

Comment (peppy) [reply osu.Server.Spectator.Tests/MultiplayerFlowTests.cs:517]: I've fixed this and checked tests pass locally. Will add CI in a separate PR.

**Pull request merged**